File: dp/p1277_count_square_submatrices/p1277_count_square_submatrices.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        
        len_n = len(matrix)
        len_m = len(matrix[0])

        subsquares_represented = [[0 for _ in range(len_m)] for _ in range(len_n)]

        for r in range(len_n):
            for c in range(len_m):

                #case that matrix[r][c] == 0
                #then that matrix element cannot represent any subsquares, s_r[r][c] = 0
                if matrix[r][c] == 0:
                    subsquares_represented[r][c] = 0
                
                #border element, can never represent more than 1 square -> s_r[r][c] = 1
                elif matrix[r][c] == 1 and (r == 0 or c == 0):
                    subsquares_represented[r][c] = 1
                    
                #case that matrix[r][c] == 1
                #then matr. el. represents the minimum number of squares of its up and left neighbors + 1
                else:
                    up     = subsquares_represented[r-1][c]
                    left   = subsquares_represented[r][c-1]
                    upleft = subsquares_represented[r-1][c-1]
                    least_squares = min(up, left, upleft)

                    subsquares_represented[r][c] = least_squares + 1
        
        return sum([sum(x) for x in subsquares_represented])

                

File: dp/p1277_count_square_submatrices/p1277_count_square_submatrices.rs
impl Solution {
    pub fn count_squares(matrix: Vec<Vec<i32>>) -> i32 {
        
        let len_n: usize = matrix.len();
        let len_m        = matrix[0].len();

        let mut represented_squares = vec![vec![0; len_m]; len_n];

        // loop over each element
        for r in 0..len_n {
            for c in 0..len_m {

            
                match matrix[r][c] {
                    0 => {/*noop, rep_sq[r][c] already 0*/}
                    1 => {
                        //edge case
                        if (r == 0 || c == 0) {
                            represented_squares[r][c] = 1;
                        }
                        //determine whether a larger square is formed by
                        //finding the minimum value of adj. elements
                        else {
                            let up     = represented_squares[r-1][c  ];
                            let left   = represented_squares[r  ][c-1];
                            let upleft = represented_squares[r-1][c-1];

                            represented_squares[r][c] = vec![up, left, upleft]
                            .iter()
                            .min()
                            .unwrap() + 1;
                        }
                    }
                    _ => println!("bad input."),
                }//end match
            }
        }//end iteration over matrix

        let ult_sum = represented_squares.iter().fold(0, |accum, row| {
            accum + row.iter().fold(0, |row_accum, &el| row_accum + el)
        });

        return ult_sum;

    }
}
